### PR TITLE
Update 2024-03

### PR DESCRIPTION
**Pull request by selenography**
Update the IPDL parser to handle various changes made to the language over the past several months.

_N.B.:_ I'm not sure if any downstream tooling is expecting `impl Display for Location` to be space-separated, so  40165b4c0b3a778ab164f2acdf82c4190c511c6c may need to be excluded. Nothing herein is dependent on it.

### CODE REVIEW

**Comment by selenography** [comment]
Apologies; I've just realized that this needs additional tests written (and the current tests need to be adjusted).

**Comment by selenography** [comment]
Never mind -- `cargo test` doesn't pass on master anyway! 🙃

I've included a few fixes that improve the situation, but updating and fixing all the tests is a little more work than I'm prepared to do at the moment — this was just a "while I'm waiting for other things to compile" project.

**Review by amccreight — CHANGES_REQUESTED** [review]
Like I mentioned in the IPC channel, it looks like the right fix here is to delete the entire check. https://hg.mozilla.org/mozilla-central/rev/bdcde8c620b43cde3941617f816a05cd2e6b716d

File: src/type_check.rs [hunk 1]
```diff
@@ -179,6 +187,7 @@ impl IPDLType {
 
         match itype {
             IPDLType::ActorType(_, _) => (),
+            IPDLType::ImportedCxxType(_, Lifetime::RefCounted, _, _) => (),
```
Comment (amccreight) [inline src/type_check.rs:190]: (Note: this came from bug 1607634, unlike the rest of the patch which came from bug 1812156)

File: tests/smoke_test.rs [hunk 2]
```diff
@@ -72,12 +72,14 @@ fn test_files(test_file_path: &str, should_pass: bool) {
                         actual_error
                     );
                     for expected_error in file_expected_error(&entry.path()) {
-                        assert!(
-                            actual_error.find(&expected_error).is_some(),
-                            "Expected \"{}\" in \"{}\"",
-                            expected_error,
-                            actual_error
-                        );
+                        if !actual_error.find(&expected_error).is_some() {
```
[the remark this answers is not in the document]
Comment (amccreight) [reply tests/smoke_test.rs:75]: Ok, I implemented lexer error translation locally. Thanks for bringing this up.

**Comment by amccreight** [comment]
I moved most of this pull request, with some modifications, into a new pull request. The last patch is still missing, but hopefully get a chance to look that over soon. The new one is here: https://github.com/mozsearch/ipdl_parser/pull/16

File: src/type_check.rs (hunk 1 above)
Comment (amccreight) [reply src/type_check.rs:190]: In my pull request, I split this change out into a separate diff, and implemented the rest of bug 1812156.

**Comment by amccreight** [comment]
I've now landed all of the patches here (or rewritten them). cargo check should pass now. Thanks for your contributions!

**Pull request closed without merging**